source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,145 @@
+ */
+class BufferedIoSocketHandleImpl : public IoHandle,
+                                   public ReadWritable,
+                                   protected Logger::Loggable<Logger::Id::io> {
+public:
+  BufferedIoSocketHandleImpl();
+
+  ~BufferedIoSocketHandleImpl() override { ASSERT(closed_); }
+
+  // IoHandle
+  os_fd_t fdDoNotUse() const override { return INVALID_SOCKET; }
+  Api::IoCallUint64Result close() override;
+  bool isOpen() const override;
+  Api::IoCallUint64Result readv(uint64_t max_length, Buffer::RawSlice* slices,
+                                uint64_t num_slice) override;
+  Api::IoCallUint64Result read(Buffer::Instance& buffer, uint64_t max_length) override;
+  Api::IoCallUint64Result writev(const Buffer::RawSlice* slices, uint64_t num_slice) override;
+  Api::IoCallUint64Result write(Buffer::Instance& buffer) override;
+  Api::IoCallUint64Result sendmsg(const Buffer::RawSlice* slices, uint64_t num_slice, int flags,
+                                  const Address::Ip* self_ip,
+                                  const Address::Instance& peer_address) override;
+  Api::IoCallUint64Result recvmsg(Buffer::RawSlice* slices, const uint64_t num_slice,
+                                  uint32_t self_port, RecvMsgOutput& output) override;
+  Api::IoCallUint64Result recvmmsg(RawSliceArrays& slices, uint32_t self_port,
+                                   RecvMsgOutput& output) override;
+  Api::IoCallUint64Result recv(void* buffer, size_t length, int flags) override;
+  bool supportsMmsg() const override;
+  bool supportsUdpGro() const override;
+  Api::SysCallIntResult bind(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult listen(int backlog) override;
+  IoHandlePtr accept(struct sockaddr* addr, socklen_t* addrlen) override;
+  Api::SysCallIntResult connect(Address::InstanceConstSharedPtr address) override;
+  Api::SysCallIntResult setOption(int level, int optname, const void* optval,
+                                  socklen_t optlen) override;
+  Api::SysCallIntResult getOption(int level, int optname, void* optval, socklen_t* optlen) override;
+  Api::SysCallIntResult setBlocking(bool blocking) override;
+  absl::optional<int> domain() override;
+  Address::InstanceConstSharedPtr localAddress() override;
+  Address::InstanceConstSharedPtr peerAddress() override;
+
+  void initializeFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                           Event::FileTriggerType trigger, uint32_t events) override;
+  IoHandlePtr duplicate() override;
+  void activateFileEvents(uint32_t events) override;
+  void enableFileEvents(uint32_t events) override;
+  void resetFileEvents() override;
+
+  Api::SysCallIntResult shutdown(int how) override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override { return absl::nullopt; }
+
+  Buffer::WatermarkBuffer& getBufferForTest() { return pending_received_data_; }
```

Comment:
Which buffer is this?

Reply:
The owned one. Will add comment

Reply:
Method erased.
Execute dynamic cast in test case.